### ISSUE TITLE
Slate plugin freeze and high CPU usage with react 18

### ISSUE THREAD

**Issue by fauh45** [react-page/react-page#1182]
**Describe the bug**
When I add a Text cell on react page editor, and I start to type at first there are no words coming up at all. Then sometimes it shows, then get stuck, the editor just freezes up. When I check the Brave browser task manager, I see that the page are using a lot of CPU (100-ish). I manage to get a recording of the Performance (snapshot using the performance tab of Chrome/Brave DevTools) that can be downloaded from this [link](https://drive.google.com/file/d/1RR7iHS75jij11jc6T2wawrBDso6w2mFW/view?usp=sharing).

![Perfomance Recording Screencapture](https://user-images.githubusercontent.com/7534613/167858863-9bbdab1c-bd71-420c-a905-a99e47fa7a6b.png)

From the screenshot of the performance recording itself, you can see how there are really deep calls, and repeated calls at the end. Is this normal? 

**To Reproduce**
1. Create React Page Editor for React Admin field on NextJS

The plugin used itself are,

```json
{
    "@react-page/editor": "^4.9.0",
    "@react-page/plugins-background": "^4.9.0",
    "@react-page/plugins-divider": "^4.9.0",
    "@react-page/plugins-html5-video": "^4.9.0",
    "@react-page/plugins-image": "^4.9.0",
    "@react-page/plugins-slate": "^4.9.0",
    "@react-page/plugins-spacer": "^4.9.0",
    "@react-page/plugins-video": "^4.9.0",
}
```

2.  Add New Text Cell
3. Try to edit, and it should be freezing, and it'll have high CPU usage

**Expected behavior**
The editor should behave just like how it is the demonstration website, smooth.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: elementary OS 6.1 Jólnir
 - Browser Brave
 - Version 4.9.0

**Additional context**
Here are the full `package.json`

```json
{
  "name": "*******",
  "version": "*.*.*",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "postinstall": "patch-package"
  },
  "dependencies": {
    "@prisma/client": "3.13.0",
    "@react-page/editor": "^4.9.0",
    "@react-page/plugins-background": "^4.9.0",
    "@react-page/plugins-divider": "^4.9.0",
    "@react-page/plugins-html5-video": "^4.9.0",
    "@react-page/plugins-image": "^4.9.0",
    "@react-page/plugins-slate": "^4.9.0",
    "@react-page/plugins-spacer": "^4.9.0",
    "@react-page/plugins-video": "^4.9.0",
    "@react-page/react-admin": "^4.9.0",
    "next": "12.1.5",
    "ra-data-simple-prisma": "^0.2.11",
    "react": "18.1.0",
    "react-admin": "^4.0.3",
    "react-dom": "18.1.0",
    "react-query": "^3.39.0"
  },
  "devDependencies": {
    "@types/node": "17.0.29",
    "@types/react": "18.0.8",
    "@types/react-dom": "18.0.0",
    "eslint": "8.14.0",
    "eslint-config-next": "12.1.5",
    "patch-package": "^6.4.7",
    "postinstall-postinstall": "^2.1.0",
    "prisma": "^3.13.0",
    "typescript": "4.6.3"
  }
}
```

For the field editor component implemented like this,

```tsx
import React from "react";

import type { EditorProps } from "@react-page/editor";
import Editor from "@react-page/editor";
import "@react-page/editor/lib/index.css";

import { Paper } from "@material-ui/core";
import { Labeled, useInput } from "react-admin";

const isEmpty = (obj: Object) => {
  for (var _ in obj) return false;
  return true;
};

export type RaReactPageInputProps = {
  label?: string;
  source: string;
  style?: React.CSSProperties;
} & EditorProps;
const RaReactPageInput: React.FC<RaReactPageInputProps> = ({
  label = "Content",
  source,
  style,
  ...editorProps
}) => {
  const {
    field: { value, onChange },
  } = useInput({ source });

  console.log("Value");
  console.log(value);

  return (
    <Labeled label={label} source={source} fullWidth>
      <>
        <Paper
          elevation={5}
          style={{
            overflow: "visible",
            padding: 16,
            marginRight: 64,

            ...style,
          }}
        >
          <Editor
            value={isEmpty(value) ? null : value}
            onChange={(v) => {
              console.log("On change");
              console.log(v);

              onChange(v);
            }}
            {...editorProps}
          />
        </Paper>
      </>
    </Labeled>
  );
};

export default RaReactPageInput;
```

And the complete cellPlugin file
```ts
import type { CellPlugin } from "@react-page/editor";

// The background plugin
import background, { ModeEnum } from "@react-page/plugins-background";
// import css as well. currently, we caannot do this here in the demo project and have moved that to _app.tsx
// see https://github.com/vercel/next.js/issues/19717
import "@react-page/plugins-background/lib/index.css";

// The divider plugin
import divider from "@react-page/plugins-divider";

// The html5-video plugin
import html5video from "@react-page/plugins-html5-video";
import "@react-page/plugins-html5-video/lib/index.css";

// The image plugin
import type { ImageUploadType } from "@react-page/plugins-image";
import { imagePlugin } from "@react-page/plugins-image";
import "@react-page/plugins-image/lib/index.css";

// The spacer plugin
import spacer from "@react-page/plugins-spacer";
import "@react-page/plugins-spacer/lib/index.css";

// The video plugin
import video from "@react-page/plugins-video";
import "@react-page/plugins-video/lib/index.css";

import slate from "@react-page/plugins-slate";
import "@react-page/plugins-slate/lib/index.css";

const fakeImageUploadService: (url: string) => ImageUploadType =
  (defaultUrl) => (file, reportProgress) => {
    return new Promise((resolve, reject) => {
      let counter = 0;
      const interval = setInterval(() => {
        counter++;
        reportProgress(counter * 10);
        if (counter > 9) {
          clearInterval(interval);
          alert(
            "Image has not actually been uploaded to a server. Check documentation for information on how to provide your own upload function."
          );
          resolve({ url: URL.createObjectURL(file) });
        }
      }, 100);
    });
  };

// Define which plugins we want to use.
export const cellPlugins: CellPlugin[] = [
  // @ts-expect-error
  slate((def) => ({
    ...def,
    id: "cms-jtk-slate",
    plugins: {
      headings: def.plugins.headings,
      alignment: def.plugins.alignment,
      emphasize: def.plugins.emphasize,
      paragraphs: def.plugins.paragraphs,
      lists: def.plugins.lists,
      link: def.plugins.link,
    },
  })),
  // @ts-expect-error
  spacer,
  // @ts-expect-error
  imagePlugin({ imageUpload: fakeImageUploadService("/images/react.png") }),
  // @ts-expect-error
  video,
  divider,
  // @ts-expect-error
  html5video,
  // @ts-expect-error
  background({
    imageUpload: fakeImageUploadService("/images/sea-bg.jpg"),
    enabledModes:
      ModeEnum.COLOR_MODE_FLAG |
      ModeEnum.IMAGE_MODE_FLAG |
      ModeEnum.GRADIENT_MODE_FLAG,
  }),
];
```

**Comment by macrozone** [comment]
hi @fauh45 



- Do you use our react-admin component? https://react-page.github.io/docs/#/integration-react-admin 
It has not yet been tested with version 4 of react-admin though (i am a bit behind of integrating that into my own projects)
- high cpu usage hint to continuous re-rendering. We rely on react's useEffect, useCallback and useMemo 's dependencies in many areas. This makes sure that if you change props  that the editor reacts properly. Downside is that you have to be careful with stuff that is "new" in each rerender. Most common pitfall is the usage of anonymous functions. E.g. in your copy of our `RaReactPageInputProps`-component you have `onChange` as an anonymous function because you log out the value. This anonymous function will be "new" in each rerender and this kills the memoization. Does your problem persist if you remove that?
- On a side note: I see that you are using react-admin with prisma (ra-data-simple-prisma). I and my company also use that quite often and we maintain a more fleshed out version of the data-provider: https://github.com/panter/ra-data-prisma. I recommend to give it a try in particular because ra-data-simple-prisma is not maintained.

**Comment by macrozone** [comment]
@fauh45 i just tried to naivly update react-admin in the example (and the package to version 4) and did not experience this problem yet, i opened the progress here https://github.com/react-page/react-page/pull/1183. The only problem is some material-ui glitches (which is another topic...)

i thinking about how we can support both versions of react-admin (3 and 4) in this integration. The difference is very small so far (just the slightly different result of `useInput`)

**Comment by fauh45** [comment]
@macrozone 

A little update again, I've tried to clone the example and try to run it. But whenever I tried to edit the text, the same thing would happened again. Also, I'm not able to get the React DevTools Profiler to run properly, as every time the editor causes the tab to freeze, the profiler would freeze too, not allowing me to stop nor start the profiling.

**Comment by macrozone** [comment]
> @macrozone
> 
> A little update again, I've tried to clone the example and try to run it. But whenever I tried to edit the text, the same thing would happened again. Also, I'm not able to get the React DevTools Profiler to run properly, as every time the editor causes the tab to freeze, the profiler would freeze too, not allowing me to stop nor start the profiling.

did you clone it directly without any change? Which browser (brave i guess?) and os? Do you experience the same thing with a different browser?

**Comment by fauh45** [comment]
Ah right, I did change some part of it. Mainly changing the react version and to use react admin 4. The full code can be seen in my [repository](https://github.com/fauh45/react-page-examples).

**Comment by macrozone** [comment]
> Ah right, I did change some part of it. Mainly changing the react version and to use react admin 4. The full code can be seen in my [repository](https://github.com/fauh45/react-page-examples).

ok, i checked it out. It has nothing to do with react-admin,

the problem is that you use react 18, which unfortunatly does not work yet https://github.com/react-page/react-page/pull/1173

not sure why there is an endless loop though, i guess that should be fixable

**Comment by fauh45** [comment]
> the problem is that you use react 18, which unfortunatly does not work yet #1173

I did change the the settings as #1173 does tho, and for the most part it does works great. Just with the slate editor, that's all.

> not sure why there is an endless loop though, i guess that should be fixable

That one I have absolutely no idea on.

**Comment by fauh45** [comment]
@macrozone 

*Another update*

I tried using react 17 and try to match the dependencies as close as the example in StackBlitz, and it works, no more freezing nor high CPU usage. Now I wonder why doesn't it work with RA 4 and React 18.

You can see the demonstration in this [StackBlitz](https://stackblitz.com/edit/github-c7wrrh-yncwp7?file=pages/index.tsx).

**Comment by macrozone** [comment]
@fauh45  i think RA4 i not the culprit, its react 18.

in the meantime, i updated slate to the newest version and had to alter the integration a bit. I did not yet try it with react 18, i don't even know whether slate itself works it it, but maybe we are lucky. 

You can give 4.10.0-beta.4 a try (will also test it soon with react 18)

**Comment by fauh45** [comment]
We ended up using React 17 as it was not detrimental to us. But if I go free time to come, for sure I'll be trying out the changes!

**Comment by gerom76** [comment]
I have provided the [pull request ](https://github.com/react-page/react-page/pull/1219) with the workaround. It works fine under React 18 but I had to disable some features (described in this PR). I believe the indicated features pinpoint the source of the problem. Any comments and hints on how to recover them would be appreciated.

**Comment by macrozone** [comment]
fixed in 5.1.4
